CMakeLists.txt

```diff
@@ -33,6 +33,7 @@ option(UR_USE_UBSAN "enable UndefinedBehaviorSanitizer" OFF)
 option(UR_USE_MSAN "enable MemorySanitizer" OFF)
 option(UR_USE_TSAN "enable ThreadSanitizer" OFF)
 option(UR_ENABLE_TRACING "enable api tracing through xpti" OFF)
+option(UR_ENABLE_SANITIZER "enable device sanitizer" OFF)
```

Comment:
Since this is disabled by default, it will not build in the CI. This patch also doesn't contain any tests.

For CI, I think we should just always have `UR_ENABLE_SANITIZER` set to `ON`. See how tracing does that here:
https://github.com/oneapi-src/unified-runtime/blob/main/.github/workflows/cmake.yml#L72

For tests, it would be useful to have something that verifies this functionality. From reading the code, it seems this currently works only on PVC? We currently don't have PVC CI runners, but we are hoping to add some soon.

Reply:
Ok, I will add some tests later.

> From reading the code, it seems this currently works only on PVC? We currently don't have PVC CI runners, but we are hoping to add some soon.

Yes, it only works on PVC now. Support for CPU will be added in an upcoming patch.

Reply:
I changed UR_ENABLE_SANITIZER to be enabled by default, because as explained in documentation, this layer should only be enabled by SYCL/OpenMP runtime. There will be no overheads when the user program is not compiled with the flag "-fsanitizer=address".

Reply:
> I changed UR_ENABLE_SANITIZER to be enabled by default, because as explained in the documentation, this layer should only be enabled by SYCL/OpenMP runtime. There will be no overheads when the user program is not compiled with the flag "-fsanitizer=address".

Makes sense.

> I found it's hard to add basic tests now, because UR_LAYER_ASAN needs some virtual memory APIs to be active. If not, it will crash on calling a null pointer.
> Hope this PR https://github.com/oneapi-src/unified-runtime/pull/939 can be landed ASAP.

Let's wait with tests until that is merged then. Thanks.

Reply:
So far I've only reviewed the generic/UR-specific parts of this PR, without delving too deeply into the specifics of how the sanitizer works. I was hoping someone familiar with this would review it. I'll take a look today.

As for tests, #939 is ready to merge soon, so it should be possible to add something soon. But unfortunately PVC runners aren't going to be available any time soon. For now, we can merge it without tests since this is fairly self-contained, but please create a github issue to create them.

Reply:
> or now, we can merge it without tests since this is fairly self-contained, but please file a github issue to create them.

Ok, I have removed the tests now, and opened a new issue for tracking this.
https://github.com/oneapi-src/unified-runtime/issues/1175

Thanks!